### PR TITLE
Reconnect Process: Partial Reconnect

### PR DESCRIPTION
**Pull request by sergeymitr**
~Requires `D46648-code` and `D47087-code`  patches to be applied.~ Patches deployed.

#### Changes proposed in this Pull Request:
- Blog and user tokens get validated via the `jetpack-token-health` WP.com endpoint.
- If the user token is invalid, it gets refreshed.
- If the blog token is invalid, it gets refreshed via the  `jetpack-refresh-blog-token` WP.com endpoint.

#### Jetpack product discussion
`p9dueE-1CV-p2`

#### Does this pull request change what data or activity we track or use?
Yes.
It adds three new events for the "Restore Connection" process:
- `restore_connection_reconnect` (full reconnect started)
- `restore_connection_refresh_blog_token` (blog token refresh initiated)
- `restore_connection_refresh_user_token` (user token refresh initiated)
Any of these events can only be triggered after Jetpack has already been connected, therefore TOS have been accepted.
Additionally, there's [TOS check](https://github.com/Automattic/jetpack/blob/master/packages/tracking/src/class-tracking.php#L107) in the code that doesn't allow the events to be tracked if TOS wasn't accepted.

#### Testing instructions:
1. Check out the branch, install the "Debug Tools", activate the "Broken Token".
2. Break the blog token.
3. Go to Jetpack Dashboard and confirm that you see the error notice.
4. Click "Restore Connection" button.
5. The dashboard should reload, the blog token should be restored. You *should not* have to click on the "Approve" button.
7. Break the user token, save somewhere existing blog token.
8. Add following code into your theme's `functions.php`:
```php
add_filter( 'react_connection_errors_initial_state', function( $errors ) {
	$errors[] = array(
		'message' => __('Connection Error'),
		'action' => 'reconnect',
		'code' => 'invalid_blog_token',
	);

	return $errors;
} );
```
9. Reload the Jetpack Dashboard, click "Restore Connection" in the error notice.
10. The authorization iframe should appear below the error notice, click "Approve".
11. Jetpack Dashboard should reload, the broken user token should be restored. The blog token _should not_ change.
12. **Do not** break any tokens, but open the "Broken Token" tab and save the current tokens somewhere.
13. Go to the Jetpack Dashboard, click "Restore Connection". Go through the authorization process.
14. Open another "Broken Token" tab and confirm that both blog and user tokens have changed.

#### Proposed changelog entry for your changes:
* Improved restoring connection to WordPress.com in case of connection errors.

### CODE REVIEW

File: packages/connection/src/class-client.php [hunk 1]
```diff
@@ -167,6 +167,57 @@ public static function remote_request( $args, $body = null ) {
 		return self::_wp_remote_request( $url, $request );
 	}
 
+	/**
+	 * Fetches a signed token.
+	 *
+	 * @param string $token the token.
+	 * @return string a signed token
+	 */
+	public static function get_signed_token( $token ) {
```
[the remark this answers is not in the document]
Comment (sergeymitr) [reply packages/connection/src/class-client.php]: Done 👍

File: packages/connection/src/class-manager.php [hunk 2]
```diff
@@ -1473,6 +1476,95 @@ public function reconnect() {
...
+
+		return (bool) count( $invalid_tokens );
+	}
+
+	/**
+	 * Perform the API request to validate the blog and user tokens.
+	 *
+	 * @return array|false The API response: `array( 'blog_token_is_healthy' => true|false, 'user_token_is_healthy' => true|false )`.
+	 */
+	private function validate_tokens() {
```
[the remark this answers is not in the document]
Comment (sergeymitr) [reply packages/connection/src/class-manager.php]: Done, good catch 👍

**Comment by leogermani** [comment]
Looking here looking here https://github.com/Automattic/jetpack/pull/16577/files#diff-159e98e335d2243f72b8d39fab156d17R2581 I see you use the Primary User token to refresh the blog token. This means this endpoint expects the primary user token to be healthy.

On the other hand, here https://github.com/Automattic/jetpack/pull/16577/files#diff-159e98e335d2243f72b8d39fab156d17R1556 you validate the token for the current user, and not the primary user.

So if I get this right, there's no guarantee that we have a valid primary user token to use.

What do you think of the following approach?

* only perform any of this if current user is an admin (or have the right capability to make the connection)
* use the current user id to refresh the blog token, which will effectively switch the blog owner to this user, in case it was broken

**Comment by fgiannar** [comment]
Hey @leogermani ,

> So if I get this right, there's no guarantee that we have a valid primary user token to use.

Great observation, you beat me to it, as I was just about to raise a question over this!

So far the wpcom endpoint only checks `is_jetpack_authorized_for_user` in its permission callback. Should we make that stricter? Would it maybe make sense for any user with a connected wpcom account to refresh the blog token?
I'm not very familiar with our corresponding capabilities yet, so please let me know in order to adjust this on both ends.

> use the current user id to refresh the blog token, which will effectively switch the blog owner to this user, in case it was broken

Where does that happen and is it indeed intended?

**Comment by leogermani** [comment]
> Hey @leogermani ,
> 
> > So if I get this right, there's no guarantee that we have a valid primary user token to use.
> 
> Great observation, you beat me to it, as I was just about to raise a question over this!
> 
> So far the wpcom endpoint only checks `is_jetpack_authorized_for_user` in its permission callback. Should we make that stricter? Would it maybe make sense for any user with a connected wpcom account to refresh the blog token?
> I'm not very familiar with our corresponding capabilities yet, so please let me know in order to adjust this on both ends.
> 

The capability is `jetpack_connect` (https://github.com/Automattic/jetpack/blob/update%2Fdebug-tools-standalone/packages/connection/src/class-manager.php#L1112) which by default falls back to `manage_options`.

On the wpcom side maybe check if the user is `administrator`?

> > use the current user id to refresh the blog token, which will effectively switch the blog owner to this user, in case it was broken
> 
> Where does that happen and is it indeed intended?

I'm not sure it already happens automatically. But it should :) 

The idea is that this Reconnect process would add fix connections where you have a "lost" primary user.

**Comment by sergeymitr** [comment]
> The capability is jetpack_connect

@leogermani we use `jetpack_disconnect` capability for the endpoint.
But that's a good point, we should use both.

**Pull request merged**